source/common/common/envoy_defaults.cc

```diff
@@ -0,0 +1,19 @@
+#include "source/common/common/envoy_defaults.h"
+
+const Envoy::DefaultsProfile& getDefaultsProfile() {
+  if (Envoy::DefaultsProfileSingleton::getExisting()) {
+    return Envoy::DefaultsProfileSingleton::get();
+  }
+
+  static Envoy::DefaultsProfile no_singleton_profile_;
+  return no_singleton_profile_;
+};
+
+const Envoy::DefaultsProfile& Envoy::DefaultsProfile::get() {
+  if (Envoy::DefaultsProfileSingleton::getExisting()) {
+    return Envoy::DefaultsProfileSingleton::get();
+  }
+
+  static Envoy::DefaultsProfile no_singleton_profile_;
+  return no_singleton_profile_;
```

Comment:
I think you might be able to use the `CONSTRUCT_ON_FIRST_USE` macro that does something similar.